src/audio/copier/copier_ipcgtw.c

```diff
@@ -250,3 +223,94 @@ void ipcgtw_zephyr_reset(struct comp_dev *dev)
 		comp_warn(dev, "ipcgtw_reset(): no buffer found");
 	}
 }
+
+int copier_ipcgtw_create(struct comp_dev *parent_dev, struct copier_data *cd,
+			 struct comp_ipc_config *config,
+			 const struct ipc4_copier_module_cfg *copier)
+{
+	struct ipcgtw_data *ipcgtw_data;
+	const struct ipc4_copier_gateway_cfg *gtw_cfg;
+	const struct ipc4_ipc_gateway_config_blob *blob;
+	int ret;
+
+	cd->ipc_gtw = true;
+
+	/* create_endpoint_buffer() uses this value to choose between input and
+	 * output formats in copier config to setup buffer. For this purpose
+	 * IPC gateway should be handled similarly as host gateway.
+	 */
+	config->type = SOF_COMP_HOST;
+
+	ret = create_endpoint_buffer(parent_dev, cd, config, copier, ipc4_gtw_none, false, 0);
+	if (ret < 0)
+		return ret;
+
+	gtw_cfg = &copier->gtw_cfg;
+	if (!gtw_cfg->config_length) {
+		comp_err(parent_dev, "ipcgtw_create(): empty ipc4_gateway_config_data");
+		ret = -EINVAL;
+		goto e_buf;
+	}
+
+	ipcgtw_data = rzalloc(SOF_MEM_ZONE_RUNTIME, 0, SOF_MEM_CAPS_RAM, sizeof(*ipcgtw_data));
+	if (!ipcgtw_data) {
+		ret = -ENOMEM;
+		goto e_buf;
+	}
+
+	ipcgtw_data->node_id = gtw_cfg->node_id;
+	ipcgtw_data->dev = parent_dev;
+
+	blob = (const struct ipc4_ipc_gateway_config_blob *)
+		((const struct ipc4_gateway_config_data *)gtw_cfg->config_data)->config_blob;
+
+	/* Endpoint buffer is created in copier with size specified in copier config. That buffer
+	 * will be resized to size specified in IPC gateway blob later in ipcgtw_params().
+	 */
+	comp_dbg(parent_dev, "ipcgtw_create(): buffer_size: %u", blob->buffer_size);
+	ipcgtw_data->buf_size = blob->buffer_size;
+
+	list_item_append(&ipcgtw_data->item, &ipcgtw_list_head);
```

Comment:
main benefit is just make eror handling more simple?

Reply:
yes, it's better to first fully prepare an object before making it globally accessible by e.g. adding it to a global list. Unless there are also other ways in which that object can be found. E.g. if you enable a notifier with this object as its argument and also add it to a list - then if there's a race, if there's no locking, you might need to add to the list first before registering the notifier. But better to lock to avoid such races.